src/coreclr/jit/lclmorph.cpp

```diff
@@ -678,16 +678,52 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
 
         // In general we don't know how an exposed struct field address will be used - it may be used to
         // access only that specific field or it may be used to access other fields in the same struct
-        // be using pointer/ref arithmetic. It seems reasonable to make an exception for the "this" arg
-        // of calls - it would be highly unsual for a struct member method to attempt to access memory
+        // by using pointer/ref arithmetic. It seems reasonable to make an exception for the "this" arg
+        // of calls - it would be highly unusual for a struct member method to attempt to access memory
         // beyond "this" instance. And calling struct member methods is common enough that attempting to
         // mark the entire struct as address exposed results in CQ regressions.
-        bool isThisArg = user->IsCall() && (user->AsCall()->gtCallThisArg != nullptr) &&
-                         (val.Node() == user->AsCall()->gtCallThisArg->GetNode());
+        GenTreeCall* callTree  = user->IsCall() ? user->AsCall() : nullptr;
+        bool         isThisArg = (callTree != nullptr) && (callTree->gtCallThisArg != nullptr) &&
+                         (val.Node() == callTree->gtCallThisArg->GetNode());
         bool exposeParentLcl = varDsc->lvIsStructField && !isThisArg;
 
-        m_compiler->lvaSetVarAddrExposed(exposeParentLcl ? varDsc->lvParentLcl
-                                                         : val.LclNum() DEBUGARG(AddressExposedReason::ESCAPE_ADDRESS));
+        bool hasHiddenStructArg = false;
+        if (m_compiler->opts.compJitOptimizeStructHiddenBuffer)
+        {
+            if (varTypeIsStruct(varDsc) && varDsc->lvIsTemp)
+            {
+                if ((callTree != nullptr) && callTree->HasRetBufArg() && val.Node() == callTree->gtCallArgs->GetNode())
+                {
+                    // Mark the local variable associated as being defined at this position.
+                    assert(!exposeParentLcl);
```

Comment:
I am seeing this assert fire in x86 `coreclr_tests` SPMI collection.

Reply:
Was that after any changes - because I am not hitting it with the latest collection?

Reply:
Good point -- it's hit if I remove the `lvIsTemp` check.

But actually I do not see the point of the assert - if we're not exposing a promoted field, why would we need to expose its parent?

Reply:
I would recommend marking the parent local in that case instead of (effectively) asserting that we cannot see promoted locals.

(As an aside: if we number the local as part of the call (+ fix morph not to forget to set the liveness flags), `IsHiddenBufferStructArg` should become a DEBUG-only property, which would be nice)

Reply:
But we don't hit the `assert(!exposeParentLcl )` as long as we have the check for `lvIsTemp`. So probably, when we remove that check, we can do that?

>  if we number the local as part of the call 

You mean memorize which arg is return buffer?

Reply:
> But we don't hit the `assert(!exposeParentLcl)` as long as we have the check for `lvIsTemp`.

I think of it as a coincidence. When I say that `lvIsTemp` "has no contract" I do mean just that - there is no rule or check that prevents promoting temps. So the fact we're not seeing promoted temps and "everything works" here is thus not by design and correct only implicitly. I think "implicit correctness" is bad and we should strive not to introduce dependencies on it.

> You mean memorize which arg is return buffer?

Yep (or obtain its through other means such as Jakob's loop).